### ISSUE TITLE
Can't build documentation on master

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#32703]
## 🐛 Bug

Building the docs on master on my machine gives a long list of scary error messages. The docs build that we run in the CI doesn't seem to have this problem, implying that it might be a dependency issue. However, the `katex` package we use has no versioning(!) and doesn't appear to do normal releases. In the interest of making the docs build portable, we should fix the errors.

```
/usr/lib/nodejs/katex/dist/katex.js:1364
                throw new _ParseError2.default("Expected '" + text + "', got '" + this.nextToken.text + "'", this.nextToken);
                ^

ParseError: KaTeX parse error: Expected 'EOF', got '\\' at position 33: …_{t} + g_{t+1} \̲\̲
p_{t+1} = p_{t…
    at new ParseError (/usr/lib/nodejs/katex/dist/katex.js:1181:16)
    at Parser.expect (/usr/lib/nodejs/katex/dist/katex.js:1364:23)
    at Parser.parseInput (/usr/lib/nodejs/katex/dist/katex.js:1415:18)
    at Parser.parse (/usr/lib/nodejs/katex/dist/katex.js:1401:30)
    at parseTree (/usr/lib/nodejs/katex/dist/katex.js:9501:17)
    at Object.renderToString (/usr/lib/nodejs/katex/dist/katex.js:68:40)
    at Socket.<anonymous> (/usr/lib/nodejs/katex/cli.js:30:26)
    at emitNone (events.js:111:20)
    at Socket.emit (events.js:208:7)
    at endReadableNT (_stream_readable.js:1064:12)
/usr/lib/nodejs/katex/dist/katex.js:1364
                throw new _ParseError2.default("Expected '" + text + "', got '" + this.nextToken.text + "'", this.nextToken);
                ^

ParseError: KaTeX parse error: Expected 'EOF', got '\\' at position 38: …+ lr * g_{t+1} \̲\̲
p_{t+1} = p_{t…
    at new ParseError (/usr/lib/nodejs/katex/dist/katex.js:1181:16)
    at Parser.expect (/usr/lib/nodejs/katex/dist/katex.js:1364:23)
    at Parser.parseInput (/usr/lib/nodejs/katex/dist/katex.js:1415:18)
    at Parser.parse (/usr/lib/nodejs/katex/dist/katex.js:1401:30)
    at parseTree (/usr/lib/nodejs/katex/dist/katex.js:9501:17)
    at Object.renderToString (/usr/lib/nodejs/katex/dist/katex.js:68:40)
    at Socket.<anonymous> (/usr/lib/nodejs/katex/cli.js:30:26)
    at emitNone (events.js:111:20)
    at Socket.emit (events.js:208:7)
    at endReadableNT (_stream_readable.js:1064:12)
/usr/lib/nodejs/katex/dist/katex.js:1364
                throw new _ParseError2.default("Expected '" + text + "', got '" + this.nextToken.text + "'", this.nextToken);
                ^

ParseError: KaTeX parse error: Expected 'EOF', got '\\' at position 147: … (2k+1)T_{max};\̲\̲
\eta_{t+1} = \…
    at new ParseError (/usr/lib/nodejs/katex/dist/katex.js:1181:16)
    at Parser.expect (/usr/lib/nodejs/katex/dist/katex.js:1364:23)
    at Parser.parseInput (/usr/lib/nodejs/katex/dist/katex.js:1415:18)
    at Parser.parse (/usr/lib/nodejs/katex/dist/katex.js:1401:30)
    at parseTree (/usr/lib/nodejs/katex/dist/katex.js:9501:17)
    at Object.renderToString (/usr/lib/nodejs/katex/dist/katex.js:68:40)
    at Socket.<anonymous> (/usr/lib/nodejs/katex/cli.js:30:26)
    at emitNone (events.js:111:20)
    at Socket.emit (events.js:208:7)
    at endReadableNT (_stream_readable.js:1064:12)
```

cc @ezyang @gchanan @zou3519

**Comment by zou3519** [comment]
cc @jlin27

**Comment by zou3519** [comment]
I'm marking this as high pri because everyone I've asked (sample size = 3) has run into these errors. We should figure out if these errors are benign or not.

**Comment by Enealor** [comment]
I think the error you are reporting shows up in the CI environment, but is being reported as a warning rather than an error. I found this when I built the documentation locally. The warning states
```
LaTeX-incompatible input and strict mode is set to 'warn': In LaTeX, \\ or \newline does nothing in display mode [newLineInDisplayMode]
```
This appears to be referencing the errors you have reported above, as each states `ParseError: KaTeX parse error: Expected 'EOF', got '\\'`, which is how KaTeX says that an unsupported command has been invoked.

Based on this, it seems that there are a few places in the documentation where people are trying to use `\\` to create a linebreak, which is not supported behavior. I'm still looking into where this particular documentation is. What I expect to find is that a few math display boxes are not formatted as intended, which is relatively benign.

Can you tell me which version of KaTeX you are using? I believe it is reporting errors for you rather than warnings due to using an older version than 0.11.1. I received similar errors when I installed the Ubuntu version via apt, which it turns out is on version 0.8.x.

**Comment by zou3519** [comment]
@Enealor how does one check the version of katex? `katex --version` hangs for me :/

**Comment by Enealor** [comment]
I had that same issue with my Ubuntu `apt` install of katex. If you can find where it's installed (e.g. `which katex`) you can find a changelog that might help. I don't know how to get it to stop hanging - I worked out my version in this case by looking at how I installed it.

**Comment by Enealor** [comment]
@zou3519 I opened a PR (#32911) to resolve this error and some formatting issues I noticed. Can you let me know if this also resolves it for you? It passes what I was considering a reproduction of the bug, but it would be good to check.

**Comment by kurtamohler** [comment]
I'm able to get through the documentation build using `cd docs && make html` after PR #32911 was merged. This used to fail and quit without generating outputs, so I used to have to work around it to just build the docs that I needed.

However, I'm now seeing some exceptions in the stdout, but they might not actually causing any issues or not. I'm not sure since they're not causing an exit, and I see files being output (I don't know if all the outputs are complete or not though). Here are some of the exception messages:

```
reading sources... [100%] type_info                                                                      
WARNING: autodoc: failed to import function 'init_process_group' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'init_process_group'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: init_process_group

WARNING: autodoc: failed to import class 'Backend' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'Backend'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: Backend

WARNING: autodoc: failed to import function 'get_backend' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'get_backend'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: get_backend

WARNING: autodoc: failed to import function 'get_rank' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'get_rank'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: get_rank

WARNING: autodoc: failed to import function 'get_world_size' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'get_world_size'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: get_world_size

WARNING: autodoc: failed to import function 'is_initialized' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'is_initialized'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: is_initialized

WARNING: autodoc: failed to import function 'is_mpi_available' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'is_mpi_available'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: is_mpi_available

WARNING: autodoc: failed to import function 'is_nccl_available' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'is_nccl_available'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: is_nccl_available

WARNING: autodoc: failed to import function 'new_group' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'new_group'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: new_group

WARNING: autodoc: failed to import function 'send' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'send'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: send

WARNING: autodoc: failed to import function 'recv' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'recv'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: recv

WARNING: autodoc: failed to import function 'isend' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'isend'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: isend

WARNING: autodoc: failed to import function 'irecv' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'irecv'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: irecv

WARNING: autodoc: failed to import function 'broadcast' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'broadcast'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: broadcast

WARNING: autodoc: failed to import function 'all_reduce' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'all_reduce'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: all_reduce

WARNING: autodoc: failed to import function 'reduce' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'reduce'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: reduce

WARNING: autodoc: failed to import function 'all_gather' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'all_gather'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: all_gather

WARNING: autodoc: failed to import function 'gather' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'gather'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: gather

WARNING: autodoc: failed to import function 'scatter' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'scatter'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: scatter

WARNING: autodoc: failed to import function 'barrier' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'barrier'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: barrier

WARNING: autodoc: failed to import class 'ReduceOp' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'ReduceOp'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: ReduceOp

WARNING: autodoc: failed to import function 'broadcast_multigpu' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'broadcast_multigpu'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: broadcast_multigpu

WARNING: autodoc: failed to import function 'all_reduce_multigpu' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'all_reduce_multigpu'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: all_reduce_multigpu

WARNING: autodoc: failed to import function 'reduce_multigpu' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'reduce_multigpu'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: reduce_multigpu

WARNING: autodoc: failed to import function 'all_gather_multigpu' from module 'torch.distributed'; the following exception was raised:
Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 238, in safe_getattr
    return getattr(obj, name, *defargs)
AttributeError: module 'torch.distributed' has no attribute 'all_gather_multigpu'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/importer.py", line 71, in import_object
    obj = attrgetter(obj, attrname)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 226, in get_attr
    return autodoc_attrgetter(self.env.app, obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/ext/autodoc/__init__.py", line 1575, in autodoc_attrgetter
    return safe_getattr(obj, name, *defargs)
  File "/home/kurtamohler/miniconda3/envs/pytorch-cuda-dev/lib/python3.6/site-packages/sphinx/util/inspect.py", line 254, in safe_getattr
    raise AttributeError(name)
AttributeError: all_gather_multigpu
```

**Comment by zou3519** [comment]
@kurtamohler are you building on a mac? It sounds like torch.distributed wasn't built and the docs build is looking for it.

**Comment by kurtamohler** [comment]
@zou3519 , no I'm on Linux. But you're right, I've set `USE_DISTRIBUTED=0`. Do you think the doc build could potentially take that into account?

**Comment by zou3519** [comment]
So I think this particular issue (the docs spewing scary-looking warnings) should be resolved after #32911. We can verify by doing a doc build on a machine.

@kurtamohler it would be nice for the doc build to not attempt to build torch.distributed documentation if the built binary was not built with distributed support to avoid scaring devs. In practice this is not as simple as it sounds (I am not sure how to conditionally include documentation with sphinx, perhaps we would write a sphinx extension).

**Comment by Enealor** [comment]
It looks like [that feature](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html#skipping-members) already exists. I found a [simple example](https://stackoverflow.com/a/21449475) of how to include this, although it would need to check if distributed is available.

**Comment by Enealor** [comment]
@zou3519 @kurtamohler I wanted to follow-up on this by checking if there are any other conditionally compiled submodules, and asking for some testing advice. At the moment, the change I plan on would add a check to `docs/source/conf.py` so that certain submodules are verified before their documentation is built. This would add a function such as the one below. However, I don't know if `distributed` is the submodule to be worried about. Are there others I should add to the list?
```
def autodoc_skip_member(app, what, name, obj, skip, options):
    exclusions = ('distributed', 'others?')
    if name in exclusions:
        spec = find_spec('.' + name, 'torch') # checks if submodule exists, returns None otherwise
        exclude = spec is None
    else:
        exclude = False
    return skip or exclude
```
I am also unsure of how to incorporate testing this outside of checking it works as expected with users. In the CI environment, I believe the docs are only built in an environment where every submodule is available.

**Comment by zou3519** [comment]
@Enealor one use case I had is for torchvision. Right now the pytorch doc build assumes that the torchvision docs exist; it would be nice for a developer to be able to compile the docs without it spilling warnings about torchvision not existing. I don't think there are any cases other than distributed & torchvision.

In terms of testing: it's probably not worth testing the behavior with distributed because that requires rebuilding pytorch and that is relatively expensive. However, we could have a CI job that builds the docs without torchvision and asserts that no "torchvision is missing" warnings show up.

**Comment by Enealor** [comment]
After experimentation, I found my previous suggestion fails to work as desired. (Problem: the module is imported before the function is called, and so the import errors still show up.) A solution that does work to remove the import warnings that kurtamohler shared is to use `suppress_warnings`. This means adding
```
try:
    import torchvision
    from torch.distributed import init_process_group
except ImportError:
    # Suppress autodoc import warnings.
    suppress_warnings = ['autodoc.import_object']
```
near the start of `docs/source/conf.py`. (I test using `init_process_group` as `torch.distributed` will import even when unbuilt.) I'm less worried about this having unexpected side-effects since it no longer selectively skips modules, but we can also flag doc builds as failing if `suppress_warnings` is defined or non-empty.

**Comment by mattip** [comment]
xref issue gh-38011 about failing documentation CI build if there are build warnings. 

It seems the topic of this issue has changed from "Can't build ..." to "Documentation build should not error on missing components". Does that change the priority?

**Comment by rgommers** [comment]
I suspect not @mattip. Building the docs should be easy, it's quite fragile right now. Making sure there's no errors or extra warnings without `torchvision` and with `USE_DISTRIBUTED=0` would be good. 

Now that the docs are broken up in more pages, it may also be worth checking page load time. KaTeX is ~2x faster than MathJax 3 still (see [here](https://www.intmath.com/cg5/katex-mathjax-comparison.php?processor=MathJax3)), however it's pretty weird (no releases, requires installing with `npm`, less complete LaTeX support), so if load times are okay (needs checking) then getting rid of it would make building the docs a lot easier as well. @zou3519 any thoughts on this?

**Comment by zou3519** [comment]
We decided to use KaTeX a while ago though because of the page load times were bad without it. But I agree, if load times are OK without KaTeX, then we should definitely remove it. As you noted, the KaTeX dependency is not very portable and kind of weird.

**Comment by mattip** [comment]
PR gh-41335 is in place so building docs will error if there are warnings. Can we close this? If migration away from KaTex is desired, it should be a separate issue.

**Comment by rgommers** [comment]
> PR gh-41335 is in place so building docs will error if there are warnings. Can we close this? 

Looks like this is complete indeed. Can you confirm that the doc builds now work with `torchvision` or `torch.distributed` missing?

> If migration away from KaTex is desired, it should be a separate issue.

That was my suggestion; I did some benchmarking and KaTeX is still significantly faster, so let's leave it as is I'd say.

**Comment by mattip** [comment]
> Can you confirm that the doc builds now work with `torchvision` or `torch.distributed`

On master, pytorch CI builds without `torchvision` installed. 

Building documentation without `torch.distributed` fails since many of the autofuntions on the [`docs/source/distributed.rst`](https://github.com/pytorch/pytorch/blob/v1.6.0/docs/source/distributed.rst) page fail to import without it. Should we pursue a fix like in the [comment above](https://github.com/pytorch/pytorch/issues/32703#issuecomment-589449440)? I fear someone might publish such a partial documentation build somewhere and they would silently miss out on documentation for the module's functionality.

**Comment by rgommers** [comment]
> I fear someone might publish such a partial documentation build somewhere and they would silently miss out on documentation for the module's functionality.

Yeah that is a valid concern. Actually, as long as the docs build completes but with broken links, I think having a nonzero exit status is a good thing. When you're not working on distributed, building PyTorch without it is useful - saves time, and often avoids build problems - so people will want to do this. But one should never want to publish such doc builds.

Okay, let's call this good then, closing. Thanks @mattip, all./
